dispatch/dispatch.go

```diff
@@ -216,6 +233,24 @@ func (d *Dispatcher) publish(id uuid.UUID, data interface{}) error {
 	}
 }
 
+// hasListener checks whether its worth publishing a job to a
+// limited job pool where many validation checks occur
+func (d *Dispatcher) hasListener(id uuid.UUID) (bool, error) {
+	if d == nil {
+		return false, errDispatcherNotInitialized
+	}
+	if id.IsNil() {
+		return false, errChannelIsNil
+	}
+	d.rMtx.RLock()
+	defer d.rMtx.RUnlock()
+	if d.routes == nil {
+		return false, fmt.Errorf("%w routes property unset", errDispatcherNotInitialized)
+	}
+	routes := d.routes[id]
+	return len(routes) > 0, nil
```

Comment:
I would actually rm this function and move this to publish. Then change the field in `job` struct from ID to be this route slice, if length is zero then return nil like before. That way if there is routes it doesn't have a consecutive map lookup in the relayer routine. OR you could just change the default job slice to around ~10k length to then allow it to catch up. The map lookup adds small processing time to this caller routine and compounds across a bunch of orderbooks and ticker websocket updates etc. Up to you.

Reply:
I did test your first solution which came with its own complications with closed channels and panics and to protect that would have ended up worse.

The reason I don't want to just increase the length is that it increases the impacts behind the scenes - The jobs being created and sent into channels, routes being taken up unnecessarily and potentially taking away from actual subscriptions to publish and kicking the can a little bit.

I think I've come up with an even simpler solution! 0.6316n/s impact vs the 13n/s I had initially. Let me know what you think